src/lib/optimizer/join_graph_builder.hpp

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <optional>
+#include <set>
+#include <vector>
+
+#include "join_graph.hpp"
+#include "logical_query_plan/union_node.hpp"
+
+namespace opossum {
+
+class AbstractLQPNode;
+
+/**
+ * Turns an LQP into a JoinGraph.
+ *
+ * Consider this simple LQP:
+ * [0] [Projection] x1, y1
+ *  \_[1] [Predicate] x2 <= y1
+ *     \_[2] [Cross Join]
+ *        \_[3] [MockTable] x
+ *        \_[4] [MockTable] y
+ *
+ * The JoinGraph created from it would contain two vertices (x and y), one edge (between x and y, with the predicate
+ * x2 <= y1) and one output_relation (Projection, left input side)
+ */
+class JoinGraphBuilder final {
+ public:
+  /**
+   * From an LQP, build a JoinGraph. The LQP is not modified during this process.
+   * std::nullopt is returned if the JoinGraph would be trivial (i.e., one vertex, no predicates)
```

Comment:
This seems to be different from the comment in `from_lqp` above

Reply:
Adjusted comment for from_lqp, they are similar in wording now.